src/readFile.ts

```diff
@@ -0,0 +1,19 @@
+import parseEnv from 'parse-dotenv'
+import { EnvObject } from './types'
+
+/**
+ * Read .env file and parse into an array of key-value pairs.
+ * @param filePath The full name of the .env file
+ * @returns {Promise<EnvObject>} Resolves with the array of entries.
+ */
+export async function getFileEntries(filePath: string): Promise<EnvObject> {
+  return new Promise(resolve => {
+    try {
+      const fileEntries: EnvObject = parseEnv(filePath)
+      resolve(fileEntries)
+    } catch (error) {
+      if (error instanceof Error)
+        throw new Error(`Error parsing file ${filePath}: ${error.message}`)
+    }
+  })
+}
```

Comment:
This function doesn't need to be async because it's not waiting on any promises (all of the code here is synchronous).

Suggestion:
```suggestion
export function getFileEntries(filePath: string): EnvObject {
    try {
      const fileEntries: EnvObject = parseEnv(filePath)
      return fileEntries
    } catch (error) {
      if (error instanceof Error)
        throw new Error(`Error parsing file ${filePath}: ${error.message}`)
    }
}
```

Reply:
@Jrfidellis Is there harm in this function being a promise? Here is why I am asking... 

If we go with the the suggested non-async alternative, we actually need to make some changes to it. The suggested code right now fails checks with 

```
Function lacks ending return statement and return type does not include 'undefined'.
```

So, we either need to change the return type to `EnvObject | undefined` and then propagate this type change to the caller and possibly beyond paying attention to where we need to check for undefined 

```
export function getFileEntries(filePath: string): EnvObject | undefined {
  try {
    const fileEntries: EnvObject = parseEnv(filePath)
    return fileEntries
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Error parsing file ${filePath}: ${error.message}`)
    }
  }
}
```

Or create an empty object upfront and potentially return an empty object to the caller and deal with that in the caller. 

```
export function getFileEntries(filePath: string): EnvObject {
  let fileEntries: EnvObject = {}
  try {
    fileEntries = parseEnv(filePath)
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Error parsing file ${filePath}: ${error.message}`)
    }
  }
  return fileEntries
}
```

So, by simplifying on one place, we might be making things more complex in other places. Would you recommendation be different, if the calling code in `main.ts` be re-written with a promise chain?

Below is ChatGPTs answer to the question `should promises only be used for async code?`


```
Promises are a fundamental concept in JavaScript primarily used for handling asynchronous operations, but they are not limited to async code exclusively. Promises are designed to represent the eventual completion (or failure) of an asynchronous operation, providing a way to handle its result or error once it's available.

However, promises can also be used in synchronous code to manage flow control and improve readability, especially when dealing with sequences of operations where some may be asynchronous and others synchronous.

Here are some key points about promises:

**Asynchronous Operations**: Promises are most commonly used to handle asynchronous operations such as fetching data from a server, reading files, or any operation that takes time to complete and may not immediately return a result.

**Synchronous Operations**: Promises can also be useful in synchronous code. For example, if you have a sequence of synchronous operations followed by an asynchronous one, you can use promises to handle the result of the asynchronous operation in a structured way.

**Error Handling**: Promises provide a standardized way to handle both success and failure scenarios, regardless of whether the operation is asynchronous or synchronous. This makes error handling more consistent and manageable.

**Chaining**: Promises support method chaining (`then`, `catch`, `finally`), which can improve code readability and maintainability by allowing you to write sequential operations without nesting callbacks.

In conclusion, while promises are commonly associated with asynchronous code, they are a versatile tool that can also be applied to synchronous code to improve structure, readability, and error handling. They provide a standardized way to handle the outcome of operations, whether they are immediate or take time to complete.

```

I don't know what the right answer is, or even if there is the right answer, so I am just trying to learn.

Reply:
it asks for `undefined` as the return type because of the conditional throw inside the catch block, if for some reason it falls into the catch block but does not re-throw, it implicit returns `undefined`, one quick fix is to throw an `unknown error` if the error is not an instance of `Error`

Reply:
Yes, but then the function still returns either an `EnvObject` or an `Error`, which is two different datatypes, and the caller, strictly speaking, needs to check. I was just wondering if making every function a promise and attaching `.catch` at the end of the promise chain makes error-handling easier...

Reply:
With changed error handling, the function now either returns an `EnvObject` or throws an exception. See f77cf96d07e31eaecf2cb45c6fcc4933c2c4c538